pkg/backup/abs_backend_test.go

```diff
@@ -0,0 +1,197 @@
+	"bytes"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/storage"
+	"github.com/coreos/etcd-operator/pkg/backup/abs"
+)
+
+const (
+	absIntegrationTestEnvVar = "RUN_ABS_INTEGRATION_TEST"
+	absIntegrationTestNotSet = "skipping ABS integration test due to RUN_ABS_INTEGRATION_TEST not set"
+)
+
+var (
+	accountName    = storage.StorageEmulatorAccountName
+	accountKey     = storage.StorageEmulatorAccountKey
+	DefaultBaseURL = "http://127.0.0.1:10000"
+	container      = "testcontainer"
+	prefix         = "testprefix"
+	blobContents   = "ignore"
+)
+
+func TestABSBackendContainerDoesNotExist(t *testing.T) {
+	if os.Getenv(absIntegrationTestEnvVar) != "true" {
+		t.Skip(absIntegrationTestNotSet)
+	}
+
+	storageClient, err := storage.NewClient(accountName, accountKey, DefaultBaseURL, "", false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = abs.NewFromClient(container, prefix, &storageClient)
+	if err == nil {
+		t.Fatal(err)
+	}
+	if err.Error() != "container testcontainer does not exist" {
+		t.Fatal(err)
+	}
+}
+
+func TestABSBackendGetLatest(t *testing.T) {
+	if os.Getenv(absIntegrationTestEnvVar) != "true" {
+		t.Skip(absIntegrationTestNotSet)
+	}
+
+	storageClient, err := storage.NewClient(accountName, accountKey, DefaultBaseURL, "", false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	blobServiceClient := storageClient.GetBlobService()
+
+	// Create container
+	cnt := blobServiceClient.GetContainerReference(container)
+	options := storage.CreateContainerOptions{
+		Access: storage.ContainerAccessTypePrivate,
+	}
+	_, err = cnt.CreateIfNotExists(&options)
+	if err != nil {
+		if accountName == storage.StorageEmulatorAccountName {
+			t.Fatal(err, "Create container failed: If you are running with the emulator credentials, plaase make sure you have started the azurite storage emulator.")
+		}
+		t.Fatal(err, "Create container failed")
+	}
+
+	abs, err := abs.NewFromClient(container, prefix, &storageClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ab := &absBackend{ABS: abs}
+
+	if _, err := ab.save("3.1.0", 1, bytes.NewBuffer([]byte(blobContents))); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := ab.save("3.1.1", 2, bytes.NewBuffer([]byte(blobContents))); err != nil {
+		t.Fatal(err)
+	}
+
+	// test getLatest
+	name, err := ab.getLatest()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expected := makeBackupName("3.1.1", 2)
+	if name != expected {
+		t.Errorf("lastest name = %s, want %s", name, expected)
+	}
+
+	// test total
+	totalBackups, err := ab.total()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if totalBackups != 2 {
+		t.Errorf("total backups = %v, want %v", totalBackups, 2)
+	}
+
+	// test open
+	rc, err := ab.open(name)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	b, err := ioutil.ReadAll(rc)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rc.Close()
+
+	if string(b) != blobContents {
+		t.Errorf("content = %s, want %s", string(b), blobContents)
+	}
+
+	// Delete container
+	opts := storage.DeleteContainerOptions{}
```

Comment:
probably should have the container delete logic under a defer function just below the container creation logic. Therefore, if any checking condition fails, container is deleted as well.

```go
_, err = cnt.CreateIfNotExists(&options)
if err != nil {...}
defer func() {
  // Delete container
  opts := storage.DeleteContainerOptions{}
  if err := cnt.Delete(&opts); err != nil {
    t.Fatal(err)
  }
}()
```

Reply:
Thank you; amended!